# Augmentation

1. 중복 제거  
2.  

In [1]:
# read file
import pandas as pd
import numpy as np

In [2]:
dataset = pd.read_csv('../../../dataset/train/train.csv')
dataset.head()

id                                           sentence  \
0   0  〈Something〉는 조지 해리슨이 쓰고 비틀즈가 1969년 앨범 《Abbey R...   
1   1  호남이 기반인 바른미래당·대안신당·민주평화당이 우여곡절 끝에 합당해 민생당(가칭)으...   
2   2  K리그2에서 성적 1위를 달리고 있는 광주FC는 지난 26일 한국프로축구연맹으로부터...   
3   3  균일가 생활용품점 (주)아성다이소(대표 박정부)는 코로나19 바이러스로 어려움을 겪...   
4   4  1967년 프로 야구 드래프트 1순위로 요미우리 자이언츠에게 입단하면서 등번호는 8...   

                                      subject_entity  \
0  {'word': '비틀즈', 'start_idx': 24, 'end_idx': 26...   
1  {'word': '민주평화당', 'start_idx': 19, 'end_idx': ...   
2  {'word': '광주FC', 'start_idx': 21, 'end_idx': 2...   
3  {'word': '아성다이소', 'start_idx': 13, 'end_idx': ...   
4  {'word': '요미우리 자이언츠', 'start_idx': 22, 'end_id...   

                                       object_entity  \
0  {'word': '조지 해리슨', 'start_idx': 13, 'end_idx':...   
1  {'word': '대안신당', 'start_idx': 14, 'end_idx': 1...   
2  {'word': '한국프로축구연맹', 'start_idx': 34, 'end_idx...   
3  {'word': '박정부', 'start_idx': 22, 'end_idx': 24...   
4  {'word': '1967', 'start_idx': 0, 'end_idx': 3,...   

                       label     source  
0                no_relation  wikipedia  
1                no_relation   wikitree  
2              org:member_of   wikitree  
3  org:top_members/employees   wikitree  
4                no_relation  wikipedia

In [3]:
import pandas as pd

def preprocessing_dataset(dataset):
    """ 처음 불러온 csv 파일을 원하는 형태의 DataFrame으로 변경 시켜줍니다."""
    subject_word, subject_idx, subject_type = [], [], []
    object_word, object_idx, object_type = [], [], []

    for subject_entity, object_entity in zip(dataset['subject_entity'], dataset['object_entity']):
        subject_dict = eval(subject_entity)
        object_dict = eval(object_entity)

        subject_word.append(subject_dict['word'])
        subject_idx.append((subject_dict['start_idx'], subject_dict['end_idx']))
        subject_type.append(subject_dict['type'])
        object_word.append(object_dict['word'])
        object_idx.append((object_dict['start_idx'], object_dict['end_idx']))
        object_type.append(object_dict['type'])

    out_dataset = pd.DataFrame({
        'id': dataset['id'], 
        'sentence': dataset['sentence'],
        'subject_word': subject_word,
        'subject_idx': subject_idx,
        'subject_type': subject_type,
        'object_word': object_word,
        'object_idx': object_idx,
        'object_type': object_type,
        'label': dataset['label'],
        'source': dataset['source']
    })

    return out_dataset

In [4]:
train_dataset = preprocessing_dataset(dataset)
train_dataset.head()

id                                           sentence subject_word  \
0   0  〈Something〉는 조지 해리슨이 쓰고 비틀즈가 1969년 앨범 《Abbey R...          비틀즈   
1   1  호남이 기반인 바른미래당·대안신당·민주평화당이 우여곡절 끝에 합당해 민생당(가칭)으...        민주평화당   
2   2  K리그2에서 성적 1위를 달리고 있는 광주FC는 지난 26일 한국프로축구연맹으로부터...         광주FC   
3   3  균일가 생활용품점 (주)아성다이소(대표 박정부)는 코로나19 바이러스로 어려움을 겪...        아성다이소   
4   4  1967년 프로 야구 드래프트 1순위로 요미우리 자이언츠에게 입단하면서 등번호는 8...    요미우리 자이언츠   

  subject_idx subject_type object_word object_idx object_type  \
0    (24, 26)          ORG      조지 해리슨   (13, 18)         PER   
1    (19, 23)          ORG        대안신당   (14, 17)         ORG   
2    (21, 24)          ORG    한국프로축구연맹   (34, 41)         ORG   
3    (13, 17)          ORG         박정부   (22, 24)         PER   
4    (22, 30)          ORG        1967     (0, 3)         DAT   

                       label     source  
0                no_relation  wikipedia  
1                no_relation   wikitree  
2              org:member_of   wikitree  
3  org:top_members/employees   wikitree  
4                no_relation  wikipedia

In [10]:
# sentence, subject_idx, object_idx, label 이 같으면 완전 동일한 data로 간주한다.
train_dataset[train_dataset.duplicated(['sentence', 'subject_idx', 'object_idx', 'label'], keep=False) == True]

id                                           sentence subject_word  \
255      255  2010년 10월 15일 전라북도청을 방문한 맹형규 행정안전부 장관은 도내 지자체간...        전라북도청   
277      277  이날 프로그램 공개에서는 전북영산작법보존회와 김명신‧정상희의 사제동행 판소리, 관악...          강태환   
630      630  배우 마동석, 박정민, 정해인, 염정아, 최성은, 윤경호 씨가 출연하는 영화 '시동...          최성은   
1386    1386  특히, KRI 한국기록원으로부터 2년 연속 공인중개사 최다 합격자 배출 기록을 공식...        한국기록원   
1721    1721  그러나 곧 이은 총선에서 한나라당이 원내 1당을 차지했고 연정상대였던 자유민주연합은...       자유민주연합   
...      ...                                                ...          ...   
31785  31785  송천영 전 의원은 1992년 민주당을 탈당, 국민신당 창당에 참여했다가 그해 11월...        민주자유당   
31786  31786  그러나, 1940년 달라이 라마 14세의 즉위식에 참석한 중화민국은 행정원몽장위원회...   달라이 라마 14세   
31896  31896  이용빈 후보는 전)더불어민주당 광산갑 지역위원장, 전)더불어민주당 정책위원회 부의장...       더불어민주당   
32274  32274         결국 2018년 초에 국민의당은 바른미래당과 민주평화당으로 갈라지고 말았다.        바른미래당   
32282  32282  박능후 중앙재난안전대책본부 1차장(보건복지부 장관)이 중앙재난안전대책본부 정례브리핑...          박능후   

      subject_idx subject_type object_word object_idx object_type  \
255      (14, 18)          ORG       행정안전부   (29, 33)         ORG   
277      (62, 64)          PER         색소폰   (58, 60)         POH   
630      (23, 25)          PER         시동'   (44, 46)         POH   
1386      (8, 12)          ORG         KRI     (4, 6)         ORG   
1721     (39, 44)          ORG        한나라당   (14, 17)         ORG   
...           ...          ...         ...        ...         ...   
31785    (49, 53)          ORG         이회창   (72, 74)         PER   
31786    (11, 20)          PER        중화민국   (32, 35)         ORG   
31896    (10, 15)          ORG         문재인   (84, 86)         PER   
32274    (18, 22)          ORG        국민의당   (12, 15)         ORG   
32282      (0, 2)          PER    보건복지부 장관   (19, 26)         ORG   

                     label           source  
255            no_relation        wikipedia  
277            no_relation         wikitree  
630            no_relation         wikitree  
1386   org:alternate_names         wikitree  
1721           no_relation        wikipedia  
...                    ...              ...  
31785          no_relation        wikipedia  
31786          no_relation        wikipedia  
31896          no_relation         wikitree  
32274          no_relation        wikipedia  
32282            per:title  policy_briefing  

[84 rows x 10 columns]

In [15]:
# 중복을 제거해 준다.
train_dataset = train_dataset.drop_duplicates(['sentence', 'subject_idx', 'object_idx', 'label'], keep='first')

42개의 중복 건 제거됨.

In [17]:
# sentence, subject_idx, object_idx는 같은데 label이 다른 경우
# 위에서 이미 완전 중복을 제거 했기때문에 sentence, subject_idx, object_idx 로 중복이면 label만 다른 케이스
pd.set_option('display.max.colwidth', 2000)
train_dataset[train_dataset.duplicated(['sentence', 'subject_idx', 'object_idx'], keep=False) == True].sort_values(by=['sentence'], axis=0)

id  \
6749    6749   
12829  12829   
8364    8364   
32299  32299   
11511  11511   
22258  22258   
277      277   
3296    3296   
4212    4212   
25094  25094   

                                                                                                    sentence  \
6749                                       대한항공은 5일 조양호 회장의 3자녀가 보유한 싸이버스카이 주식 9만9900주 전량을 63억 원에 취득한다고 밝혔다.   
12829                                      대한항공은 5일 조양호 회장의 3자녀가 보유한 싸이버스카이 주식 9만9900주 전량을 63억 원에 취득한다고 밝혔다.   
8364                               배우 김병철 씨가 연기하는 정복동은 천리마마트를 망하게 하기 위해 여러 계획을 세우지만 반대로 하는 일마다 대박을 치는 캐릭터이다.   
32299                              배우 김병철 씨가 연기하는 정복동은 천리마마트를 망하게 하기 위해 여러 계획을 세우지만 반대로 하는 일마다 대박을 치는 캐릭터이다.   
11511                                영화 '버즈 오브 프레이'는 배트맨이 없는 고담시에서 할리퀸, 헌트리스, 블랙 카나리가 힘을 합쳐 한 소녀를 구한다는 이야기다.   
22258                                영화 '버즈 오브 프레이'는 배트맨이 없는 고담시에서 할리퀸, 헌트리스, 블랙 카나리가 힘을 합쳐 한 소녀를 구한다는 이야기다.   
277    이날 프로그램 공개에서는 전북영산작법보존회와 김명신‧정상희의 사제동행 판소리, 관악 명인 대금 원장현, 색소폰 강태환의 무대를 선보이며 올해 축제에 대한 기대감을 한껏 끌어 올렸다.   
3296   이날 프로그램 공개에서는 전북영산작법보존회와 김명신‧정상희의 사제동행 판소리, 관악 명인 대금 원장현, 색소폰 강태환의 무대를 선보이며 올해 축제에 대한 기대감을 한껏 끌어 올렸다.   
4212     한편 전라남도는 최근 확진자가 발생한 순천시와 여수시에 마스크를 각각 2만장씩 총 4만장 지원했으며, 이날 확진자가 나온 광양시에도 현장상황을 감안해 적정 물량을 지원키로 했다.   
25094    한편 전라남도는 최근 확진자가 발생한 순천시와 여수시에 마스크를 각각 2만장씩 총 4만장 지원했으며, 이날 확진자가 나온 광양시에도 현장상황을 감안해 적정 물량을 지원키로 했다.   

      subject_word subject_idx subject_type object_word object_idx  \
6749          대한항공      (0, 3)          ORG         조양호    (9, 11)   
12829         대한항공      (0, 3)          ORG         조양호    (9, 11)   
8364           정복동    (15, 17)          PER         김병철     (3, 5)   
32299          정복동    (15, 17)          PER         김병철     (3, 5)   
11511          배트맨    (16, 18)          PER         고담시   (24, 26)   
22258          배트맨    (16, 18)          PER         고담시   (24, 26)   
277            강태환    (62, 64)          PER         색소폰   (58, 60)   
3296           강태환    (62, 64)          PER         색소폰   (58, 60)   
4212          전라남도      (3, 6)          ORG         여수시   (26, 28)   
25094         전라남도      (3, 6)          ORG         여수시   (26, 28)   

      object_type                      label     source  
6749          PER                no_relation  wikipedia  
12829         PER  org:top_members/employees  wikipedia  
8364          PER                no_relation   wikitree  
32299         PER        per:alternate_names   wikitree  
11511         LOC     per:place_of_residence   wikitree  
22258         LOC                no_relation   wikitree  
277           POH                no_relation   wikitree  
3296          POH                  per:title   wikitree  
4212          LOC                org:members   wikitree  
25094         LOC  org:place_of_headquarters   wikitree

선정 기준은 가급적 type 조합이 label에 맞도록 선정하였다.

(6749), 12829 : no_relation  
8364, (32299) : per:alternate_names  
(11511), 22258 : per:place_of_residence  
277, (3296) : title  
4212, (25094) : org:place_of_headquarters  

In [28]:
train_dataset = train_dataset.drop([12829, 8364, 22258, 277, 4212])

In [29]:
len(train_dataset)

32423

[Optional]  
make mapping of type set and labels

In [138]:
label_type_mapped = {
    "no_relation": []
    ,"org:dissolved": [('ORG', 'DAT')] # 지정된 조직이 해산된 날짜 O
    ,"org:founded": [('ORG', 'DAT')] # 지정된 조직이 설립된 날짜 O
    ,"org:place_of_headquarters": [('ORG', 'LOC'), ('ORG', 'ORG'), ('ORG', 'POH')] # 지정된 조직의 본부가 있는 장소(본사위치)
    ,"org:alternate_names": [('ORG', 'ORG'), ('ORG', 'POH')] # 지정된 조직을 참조하기 위해 사무실 이름 대신 호출되는 대체 이름
    ,"org:member_of": [('ORG', 'ORG'), ('ORG', 'POH'), ('ORG', 'LOC')]
    ,"org:members": [('ORG', 'ORG'), ('ORG', 'POH'), ('ORG', 'LOC')]
    ,"org:political/religious_affiliation": [('ORG', 'ORG'), ('ORG', 'POH')]
    ,"org:product": [('ORG', 'POH')]
    ,"org:founded_by": [('ORG', 'PER')]
    ,"org:top_members/employees": [('ORG', 'PER')]
    ,"org:number_of_employees/members": [('ORG','NOH')]
    ,"per:date_of_birth": [('PER', 'DAT')]
    ,"per:date_of_death": [('PER', 'DAT')]
    ,"per:place_of_birth": [('PER', 'LOC')]
    ,"per:place_of_death": [('PER', 'LOC')]
    ,"per:place_of_residence": [('PER', 'LOC')]
    ,"per:origin": [('PER', 'LOC'), ('PER', 'ORG')]
    ,"per:employee_of": [('PER', 'ORG')]
    ,"per:schools_attended": [('PER', 'ORG')]
    ,"per:alternate_names": [('PER', 'PER')]
    ,"per:parents": [('PER', 'PER'), ('PER', 'POH')]
    ,"per:children": [('PER', 'PER'), ('PER', 'POH')]
    ,"per:siblings": [('PER', 'PER'), ('PER', 'POH')]
    ,"per:spouse": [('PER', 'PER'), ('PER', 'POH')]
    ,"per:other_family": [('PER', 'PER'), ('PER', 'POH')]
    ,"per:colleagues": [('PER', 'PER')]
    ,"per:product": [('PER', 'POH')]
    ,"per:religion": [('PER', 'ORG'), ('PER', 'POH')]
    ,"per:title": [('PER', 'POH')]
}

In [197]:
cnt = 0
# 기존 건수: 32,423
print(len(train_dataset))
for index, row in train_dataset.iterrows():
    if ((row.subject_type, row.object_type) not in label_type_mapped[row.label]) and row.label != 'no_relation':
        cnt += 1
        train_dataset.drop(index, inplace=True)

print(cnt) # 1625
print(len(train_dataset)) # 30798

32423
1625
30798


In [136]:
train_dataset[train_dataset.label == label_check].subject_type.value_counts()

PER    2099
Name: subject_type, dtype: int64

In [137]:
train_dataset[train_dataset.label == label_check].object_type.value_counts()

POH    1842
ORG     140
LOC      81
NOH      19
PER      14
DAT       3
Name: object_type, dtype: int64

In [79]:
train_dataset[(train_dataset.label == label_check) & (train_dataset.subject_type == 'PER')]

id  \
3118  3118   

                                                                                                sentence  \
3118  이후 태연, 써니, 효연, 유리, 윤아로 소녀시대의 두 번째 유닛 그룹 소녀시대-Oh!GG를 결성하였으며, 9월 5일 첫 싱글 앨범 《몰랐니(Lil' Touch)》를 발표했다.   

     subject_word subject_idx subject_type object_word object_idx object_type  \
3118         소녀시대    (23, 26)          PER  소녀시대-Oh!GG   (40, 49)         PER   

            label     source  
3118  org:members  wikipedia

In [58]:
# train_dataset.loc[(train_dataset.label == label_check) & (train_dataset.subject_type == 'PER'), 'subject_type'] = 'ORG'
# train_dataset.loc[(train_dataset.label == label_check) & (train_dataset.object_type == 'DAT'), 'object_type'] = 'LOC'
# train_dataset.loc[(train_dataset.label == label_check) & (train_dataset.object_type == 'NOH'), 'object_type'] = 'LOC'
# train_dataset.loc[(train_dataset.label == label_check) & (train_dataset.object_type == 'PER'), 'object_type'] = 'LOC'

In [134]:
train_dataset[(train_dataset.label == label_check) & (train_dataset.object_type == 'POH')].head(5)

id  \
1148    1148   
1865    1865   
2524    2524   
13584  13584   
13882  13882   

                                                                                                                                                                           sentence  \
1148                                                          그는 리처드 도킨스 (Richard Dawkins), 크리스토퍼 히친스 (Christopher Hitchens), 다니엘 데넷 (Daniel Dennett)과 함께 무신론자 4명의 기수 중 한 명으로 여겨진다.   
1865   1844년 무신론 정치 경제학자인 카를 마르크스는 그의 저서 《헤겔 법철학의 비판을 위하여, 서설》에서 '종교적 비참은 현실적 비참의 표현이자 현실적 비참에 대한 항의이다. 종교는 곤궁한 피조물의 한숨이며, 무정한 세계의 감정이고, 또 정신 없는 상태의 정신이다. 종교는 인민의 아편이다.' 라고 주장하였다.   
2524                                                                                        따라서 로베스피에르는 에베르파와 비슷하게 그리스도교를 탄압하였는데, 두 파벌은 이신론(理神論)과 무신론(無神論)이라는 점에서만 의견이 달랐다고 할 수 있다.   
13584                                                                         헨리 나우웬(Henri Jozef Machiel Nouwen, 1932년 ~ 1996년)은 네덜란드 출신의 로마 가톨릭사제이자 사목신학자이며 그리스도교 영성가로도 널리 알려져 있다.   
13882                                                                                                                1654년, 당시 잉글랜드의 군인이던 올리버 크롬웰이 청교도 혁명을 일으켜 찰스 1세를 처형하고 권력을 잡는다.   

      subject_word subject_idx subject_type object_word object_idx  \
1148           도킨스      (7, 9)          PER         무신론   (92, 94)   
1865          마르크스    (22, 25)          PER         무신론     (6, 8)   
2524        로베스피에르      (4, 9)          PER         이신론   (44, 46)   
13584       헨리 나우웬      (0, 5)          PER      로마 가톨릭   (60, 65)   
13882      올리버 크롬웰    (21, 27)          PER         청교도   (30, 32)   

      object_type         label     source  
1148          POH  per:religion  wikipedia  
1865          POH  per:religion  wikipedia  
2524          POH  per:religion  wikipedia  
13584         POH  per:religion  wikipedia  
13882         POH  per:religion  wikipedia

## 데이터 증강

In [204]:
train_dataset.label.value_counts()

no_relation                            9514
org:top_members/employees              4189
per:employee_of                        2853
org:member_of                          1849
per:title                              1842
org:alternate_names                    1258
org:place_of_headquarters              1188
per:date_of_birth                      1128
per:origin                             1088
per:alternate_names                     879
per:spouse                              787
per:colleagues                          512
per:parents                             505
org:founded                             450
org:members                             413
per:date_of_death                       409
org:product                             304
per:children                            297
per:other_family                        185
per:place_of_residence                  172
per:place_of_birth                      161
org:founded_by                          144
per:siblings                    

In [208]:
train_dataset[train_dataset['label'] == 'per:place_of_death'].head(1)

id                                             sentence subject_word  \
1044  1044  2010년 3월 29일, 최진영은 서울특별시 강남구 논현동의 자택에서 스스로 세상을 떠났다.          최진영   

     subject_idx subject_type object_word object_idx object_type  \
1044    (14, 16)          PER   서울특별시 강남구   (19, 27)         LOC   

                   label     source  
1044  per:place_of_death  wikipedia

In [206]:
additional_dataset = train_dataset.copy().head(0)
additional_dataset

Empty DataFrame
Columns: [id, sentence, subject_word, subject_idx, subject_type, object_word, object_idx, object_type, label, source]
Index: []

In [ ]:
for 